### ISSUE TITLE
Inconsistent timezone in line charts

### ISSUE THREAD

**Issue by l2dy** [allegro/turnilo#274]
I enabled `privacy.resistFingerprinting` in Firefox, so the browser timezone is always UTC. The table view displays timestamps in the browser timezone instead of the timezone selected from the dropdown in the upper right corner. X-axis of line charts are also displayed in the browser timezone.

**Comment by l2dy** [comment]
In line charts, timestamps in the tooltips and x-axis are in different time zones, which I think is definitely a bug.

**Comment by l2dy** [comment]
> In line charts, timestamps in the tooltips and x-axis are in different time zones, which I think is definitely a bug.

If the timezone selected in the dropdown is the source (Druid) timezone, and time displayed is intended to be of the browser timezone, then time in floating tooltips is incorrect, otherwise the x-axis is incorrect.

**Comment by l2dy** [comment]
The Druid query (as shown in Turnilo's dropdown menu > Display Druid query) only uses `Etc/UTC` timezone on my Turnilo instance even if I set the browser timezone apart. So I assume that Turnilo handles the source timezone independent of the timezone selected (which is expected in order to produce correct results).

I'd vote for displaying the time in the selected timezone, otherwise the selection would be meaningless, right?

**Comment by adrianmroz** [comment]
What you mean by that Druid uses etc timezone? If you looking at values in time filters, it's because they're converted to utc, but values behind them are respecting your timezone, not druid. Or is timezone passed somewhere else in query?

**Comment by l2dy** [comment]
> What you mean by that Druid uses etc timezone? If you looking at values in time filters, it's because they're converted to utc, but values behind them are respecting your timezone, not druid. Or is timezone passed somewhere else in query?

> Right, I misunderstood the Druid query. So the situation is more complex than I thought.
> 
> First, timezone affects time filters in the Druid query (converted from no timezone to utc). Do we assume that Druid clusters are always configured with `-Duser.timezone=UTC` as in http://druid.io/docs/latest/configuration/production-cluster.html (or is this irrelevant)?

I read the example in https://druid.io/docs/latest/querying/granularities.html#period-granularities. It seems that timestamps in Druid queries and responses already has timezone information, and `timeZone.granularity` in queries specifies the expected timezone used in response timestamps.

Druid knows its timezone, and we specified the timezone in queries, so the `-Duser.timezone` settings in Druid is probably irrelevant for us.

> Second, should we use browser timezone or selected timezone for this conversion?
>
> Third, time split is visualized with no timezone. Should we convert utc time from the Druid query results (We passed `"timeZone": "Etc/UTC"` in `granularity`, so time in the response should be utc, I think.) into the browser timezone or the selected timezone?

My instinct tells me that if there's a timezone selection in the UI, it should be respected throughout the interface, so I'd vote for using the selected timezone only.

**Comment by l2dy** [comment]
Here the `.toDate()` function erased the timezone information and converted the moment into browser (system) timezone, because the [`Date`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) object is *always* in the time zone of the browser. So when D3 formats this `Date` it could only use the browser timezone.

https://github.com/allegro/turnilo/blob/bd73d4c4b172603169728658d23a8607e564684d/src/client/components/line-chart-axis/line-chart-axis.tsx#L69

https://github.com/allegro/turnilo/blob/bd73d4c4b172603169728658d23a8607e564684d/src/client/components/line-chart-axis/line-chart-axis.tsx#L56-L59

**Comment by adrianmroz** [comment]
Ok, first things first, let's bring consistency to UI. I will fix timezone erasure in line-chart-axis and table.  

Second: When I select Fixed time filter and pick some start and end - it's time in my timezone (timezone selected in UI). When I send query to Druid I send it in UTC but I still attach timezone. I could send it in my timezone and attach my timezone. I don't have fine control over it, it happens inside plywood. But, when I change timezone - UI is changing - inside time filter different hours are selected - they're converted to different timezone (only converted in UI, internally it's the same hour with timezone attached). And if you look into druid query, it sends the same time. Which is correct.

And if you look at the response from Druid, it's time with UTC timezone attached, so we can convert to our selected timezone with ease.

In conclusion, In my opinion we handle timezones between ui and druid correctly. We just have few spots with incorrect formatting - namely axis on line chart and left gutter on table.

Bringing attention of @piter75 on this :)

**Comment by adrianmroz** [comment]
Closed via #317
